Core/COObject.m

```diff
@@ -1331,6 +1435,96 @@ - (void)commonDidChangeValueForProperty: (NSString *)key
 
 }
 
+/**
+ * Fast path variant of -commonDidChangeValueForProperty: that only checks
+ * the inserted/removed values
+ */
+- (void)commonDidChangeValueForProperty: (NSString *)key
+							  atIndexes: (NSIndexSet *)indexes
+							withObjects: (NSArray *)objects
+						   mutationKind: (ETCollectionMutationKind)mutationKind
+{
+	ETPropertyDescription *propertyDesc = [_entityDescription propertyDescriptionForName: key];
+	id newValue = [self serializableValueForStorageKey: key];
+	id oldValue = [self popOldCoreObjectRelationshipValueForPropertyDescription: propertyDesc];
+	
+	if (propertyDesc == nil)
+	{
+		[NSException raise: NSInvalidArgumentException
+					format: @"Property %@ is not declared in the metamodel %@ for %@",
+		 key, _entityDescription, self];
+	}
+	
+	if (![self isCoreObjectCollection: newValue])
+	{
+		// Call the slow path.
+		
+		// We rely on COUnsafeRetainedMutableArray automatic removal of duplicates,
+		// otherwise we could use the fast path for non-COPrimitiveCollection multivalues.
+		
+		[self commonDidChangeValueForProperty: key];
+		return;
+	}
+	
+	// Fast path:
+	
+	[(id <COPrimitiveCollection>)newValue setMutable: NO];
+	
+	// We must figure out which objects were added, and which were replaced or removed
+	
+	NSArray *replacedOrRemoved;
+	NSArray *addedObjects;
+	
+	if (mutationKind == ETCollectionMutationKindInsertion)
+	{
+		addedObjects = objects;
+		replacedOrRemoved = @[];
+	}
+	else if (mutationKind == ETCollectionMutationKindRemoval)
+	{
+		addedObjects = @[];
+		replacedOrRemoved = objects;
+	}
+	else if (mutationKind == ETCollectionMutationKindReplacement)
+	{
+		// FIXME: How do we know which objects were replaced if it's an unordered collection?
```

Comment:
I tried to avoid this extra snapshot when I wrote this code, but it's probably pretty hard to do, so it could be better to keep this change for another pull request I think.

-updateCachedOutgoingRelationshipsXXX relies on this snapshot and we have a check to delect badly paired -will/DidChangeValueXXX  in -popOldCoreObjectRelationshipValueForPropertyDescription:. This last method is pretty ugly, but debugging missing -will/DidChangeValueXXX is hard.

The really ugly thing is that Apple is already taking a snapshot of the old collection in -willChangeValueForKey: but gives no API to access it in -didChangeValueForKey: (-didChangeValueForProperty: in our case). It's only available when you receive the KVO notification in -observeValueForKeyPath:ofObject:change:context: (as NSKeyValueChangeOldKey). 
So I'm not sure what the solution would be. Writing some dedicated code to make these snapshots cheap or hacking into KVO private API, or something else?